File: DMT_GPIO/GPIO.py
# Importing Libraries
import sys
from datetime import datetime, date
import time
import numpy as np
import csv
import pigpio  # Library for high speed gpio

"""
@author: Jimmy van de Worp
This script receives high speed GPIO indefinitely and records the Pin Number, State, and Tick of each GPIO change detected

"""

'''

Current start-up procedure:
$ sudo pigpiod 
Go to Directory

Running procedure:

To run: ~/Directory $ python3 GPIO.py 12 16
Numbers represent the pins being checked (leaving blank will check all pins)
Defining Callback Function, GPIO = Pin No., level = state (1=HIGH, 0=LOW), tick = time (in us) since RPi bootup

'''

def procedure(GPIO, level, tick):

   '''
   This is the function that occurs when a GPIO change is detected in the pi.callback().
   It appends the GPIO Pin, State, Tick, and Time Difference since last State Change was detected to the py_data array.

    Returns:
        None 
   
   '''

   if last[GPIO] is not None:
   
      # Time difference (in us) between the current event change and the last change
   
      diff = pigpio.tickDiff(last[GPIO], tick)
   
      py_data.append([GPIO, level, tick, diff])
   
      # print("G={} l={} t={} d={}".format(GPIO, level, tick, diff)) - Debug Printing - IGNORE
   
   last[GPIO] = tick  # Resetting the new previous GPIO state and tick time


if __name__ == '__main__':

   last = [None]*32

   cb = []  # Documented Standard for library and pi.callback()
   
   py_data = [['GPIO', 'State', 'Tick', 'Diff']]  # Titles

   pi = pigpio.pi()  # Connects to Local Pi

   if not pi.connected:  # Exits code if no connection
     
      exit()

   if len(sys.argv) == 1:
   
      pins = range(0, 32)  # Total of 32 Pins
   
   else:
   
      pins = []
   
      for i in sys.argv[1:]:
   
         pins.append(int(i))  # For running code on multiple systems - IGNORE

   for pin in pins:
   
      # Calls function once event change is detected on pin g
   
      cb.append(pi.callback(pin, pigpio.EITHER_EDGE, procedure))

   # Closing Procedure

   try:

      while True:

         time.sleep(60) # Runs Indefinitely

   except KeyboardInterrupt: # Until Keyboard Interrupt

      print("\nTidying up")

      with open('time_GPIO' + str(date.today()) + '.csv', 'a+', newline='') as csvfile:  # Writing Data into .csv

         writer = csv.writer(csvfile)

         for i in py_data:

               row = [i]

               writer.writerow(row)  # Writing one row at a time

      for c in cb:

         c.cancel()

   pi.stop()

File: DMT_GPIO/GPIO.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <pigpio.h>

#define NUM_PINS 32

int last[NUM_PINS];
int cb[NUM_PINS];
int py_data[NUM_PINS][4];
int num_entries = 1; // First row reserved for header

// Defining Callback Function, GPIO = Pin No., level = state (1=HIGH, 0=LOW), tick = time (in us) since RPi bootup
void cbf(int GPIO, int level, uint32_t tick)
{
if (last[GPIO] != -1)
{
int diff = tickDiff(last[GPIO], tick); // Time difference (in us) between the current event change and the last change
py_data[num_entries][0] = GPIO;
py_data[num_entries][1] = level;
py_data[num_entries][2] = tick;
py_data[num_entries][3] = diff;
//printf("G=%d l=%d t=%d d=%d\n", GPIO, level, tick, diff);
}
last[GPIO] = tick; // Resetting the new previous GPIO state and tick time
num_entries++;
}

int main(int argc, char *argv[])
{
int num_g;
int *G;

if (argc == 1)
{
num_g = NUM_PINS;
G = (int *)malloc(num_g * sizeof(int));
for (int i = 0; i < NUM_PINS; i++)
G[i] = i;
}
else
{
num_g = argc - 1;
G = (int *)malloc(num_g * sizeof(int));
for (int i = 1; i < argc; i++)
G[i - 1] = atoi(argv[i]);
}

if (gpioInitialise() < 0)
{
printf("Failed to connect to pigpio\n");
return 1;
}

for (int i = 0; i < NUM_PINS; i++)
last[i] = -1;

for (int i = 0; i < num_g; i++)
cb[i] = gpioSetAlertFunc(G[i], cbf);

// Closing Procedure
int loop = 1;
while (loop)
{
sleep(60);
}

printf("Tidying up\n");
for (int i = 0; i < num_entries; i++)
printf("%d %d %d %d\n", py_data[i][0], py_data[i][1], py_data[i][2], py_data[i][3]);
for (int i = 0; i < num_g; i++)
gpioSetAlertFunc(G[i], NULL);

gpioTerminate();

return 0